### ISSUE TITLE
Combining setCondition and addConditionParameter can break SQL request.

### ISSUE THREAD

**Issue by Yivan** [pimcore/pimcore#1831]
Using object API to request list, can lead to malformed SQL request if setCondition is used and after addConditionParameter add some conditions.
The problem only occurs when the resulting list is empty (all the combined criterea send no result).

Here is a sample of the code:

```
$appList = new App\Listing();
if (isset($params["category"]) && $params["category"]) $appList->setCondition("category = ?", $params["category"]);
if (isset($params["isFree"])) {
	if ($params["isFree"] == "0") $appList->addConditionParam("isFree = ? OR isFree IS NULL", 0);
	if ($params["isFree"] == "1") $appList->addConditionParam("isFree = ?", 1);
}
if (isset($params["order"]) && $params["order"] == "averageRating") {
	$appList->setOrderKey("averageRating");
	$appList->setOrder("desc");
} else {
	$appList->setOrderKey("lastUpdate");
	$appList->setOrder("desc");
}
$appList->load();
```

Work ok when their is results. But as soon $appList->load() will have 0 objects, an error is sent:

```
`SQLSTATE[HY093]: Invalid parameter number: number of bound variables does not match number of tokens, query was: SELECT object_18.o_id as o_id, `object_18`.`o_type` FROM `object_18` WHERE ((category = ? AND isFree = ? AND object_18.o_type IN ('object','folder')) AND object_18.o_published = 1) ORDER BY `lastUpdate` DESC"`

file: "...\www\vendor\zendframework\zendframework1\library\Zend\Db\Statement\Pdo.php"
```

For now the solution is to replace in the code above the first `setCondition` by `addConditionParam`, and it works great even with empty resulting list.
But i think it should be corrected at the API level, as setCondition and addConditionParam could be used in combination, as indicated in the documentation:

https://www.pimcore.org/docs/latest/Objects/Working_with_PHP_API.html
```
$entries->setCondition("name LIKE ?", ["%bernie%"]); // use prepared statements! Mysqli only supports ? placeholders
// or
$entries->setCondition("name LIKE :name", ["name" => "%bernie%"]); // With PDO_Mysql you can use named parameters
// to add param to the condition
$entries->addConditionParam("city = ?", "New York", "AND"); // concatenator can be AND or OR
```

**Comment by Yivan** [comment]
Test was done on Pimcore 4 last stable build.

**Comment by Yivan** [comment]
Just for reference, it could be linked to the problem of encapsulation of addConditionParam : https://github.com/pimcore/pimcore/issues/1838

**Comment by brusch** [comment]
If this can be resolved, the hint in the docs should be removed: see #1831

**Comment by Yivan** [comment]
I should add that as I said in my first post af this issue, remplacing the `setCondition` by `addConditionParam` make it works:

Please this this screenshot:
![ok](https://user-images.githubusercontent.com/9025839/35151609-49796ad6-fd1f-11e7-8569-94a51f3065e5.png)

**Comment by Yivan** [comment]
In fact this is strange because with my code condition brand is setted to "peugeot", it should send back nothing... strange.

**Comment by weisswurstkanone** [comment]
I propose the attached patch. Reason is that the condition variables from setCondition and addCondition were not merged but overwritten

[Combining_setCondition_and_addConditionParam_can_break_SQL_request__#1831.zip](https://github.com/pimcore/pimcore/files/1646961/Combining_setCondition_and_addConditionParam_can_break_SQL_request__.1831.zip)

**Comment by Yivan** [comment]
I tested the proposed patch.
It correct well this issue, but it seems to introduce others problems.

If you apply the patch, clear the cache and reload admin you will get an infinite loading. Checking the response in dev tools give:

```
An exception occurred while executing 'SELECT `translations_admin`.* FROM `translations_admin` WHERE (language = ?)':
SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '?)' at line 1 (500 Internal Server Error)
```

If there is already an old cache and manage to arrive to the backend gui, a popin is opened with this error:

```
Status: 500 | Internal Server Error
URL: /admin/asset/tree-get-root
Message: An exception occurred while executing 'SELECT `assets`.`id`, `assets`.`type` FROM `assets` WHERE (path LIKE ?) LIMIT 35':
SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '?) LIMIT 35' at line 1
```

**Comment by weisswurstkanone** [comment]
true, getConditionVariables() also has to do this merging stuff before returning the condition variables.

please find an updated patch 
[Combining_setCondition_and_addConditionParam_can_break_SQL_request__#1831a.zip](https://github.com/pimcore/pimcore/files/1651218/Combining_setCondition_and_addConditionParam_can_break_SQL_request__.1831a.zip)

**Comment by Yivan** [comment]
@weisswurstkanone I just tested it and works great now!
With this second patch, this issue is resolved and backend seems to work ok (i have no more the errors, even if i flush all the caches).
Thanks.
